lib/ukfile/include/uk/statx.h

```diff
@@ -0,0 +1,80 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#ifndef __UKFILE_STATX_H__
+#define __UKFILE_STATX_H__ 1
+
+#include <stdint.h>
+
+/*
+ * Linux-compatible `statx` structure for use in syscalls, along with
+ * definitions of bit flags.
+ *
+ * Layout taken from statx(2) man page with paddings from musl v1.2.3.
+ * Flags taken from Linux headers v6.5.6.
+ */
+
+struct uk_statx_timestamp {
+	int64_t tv_sec;
+	uint32_t tv_nsec;
+	int32_t _reserved;
+};
+
+struct uk_statx {
+	uint32_t stx_mask;
+	uint32_t stx_blksize;
+	uint64_t stx_attributes;
+	uint32_t stx_nlink;
+	uint32_t stx_uid;
+	uint32_t stx_gid;
+	uint16_t stx_mode;
+	uint16_t _mode_reserved;
+	uint64_t stx_ino;
+	uint64_t stx_size;
+	uint64_t stx_blocks;
+	uint64_t stx_attributes_mask;
+	struct uk_statx_timestamp stx_atime, stx_btime, stx_ctime, stx_mtime;
+	uint32_t stx_rdev_major;
+	uint32_t stx_rdev_minor;
+	uint32_t stx_dev_major;
+	uint32_t stx_dev_minor;
+	uint64_t stx_mnt_id;
+	uint32_t stx_dio_mem_align;
+	uint32_t stx_dio_offset_align;
+	uint64_t _spare[12];
+};
+
+/* Bits used in stx_mask */
+#define STATX_TYPE        0x00000001U /* File type in stx_mode */
```

Comment:
To answer the first question, these are unikraft definitions that should equal Linux kernel definitions, however these are not available in libc's.

statx is a syscall interface and is not usually exposed to userspace directly. There is no `<sys/statx.h>` and it's not usually present in other `<sys/*.h>` headers either:
- glibc under `_GNU_SOURCE` will pull statx defintions into `<sys/stat.h>` by including a header from `bits/` (which are copied from kernel headers at configure-time)
- musl does not include any headers and defines its own `struct statx` in `fstatat.c`; it defines no mask bits and uses the magic value `0x7ff` in syscalls

Given these circumstances, what's the best call?

Reply:
This sounds to me then that we should prefix the definitions with `UK_` to avoid any clashes but then we can keep them in this file. Since you wrote you got them from Linux: Under which license are these defintions? I hope non-GPL or covered by the syscall note. Is it possible that  because of licensing, the bits are not defined when using musl (if by chance you know)?

Reply:
I don't know whether musl doesn't include them because of licensing issues, or because they're Linux extensions (i.e., not posix) and musl can't (yet) be bothered.
The header file itself (`include/uapi/linux/stat.h` in the kernel source tree) is under GPL2 with the syscall note:
```
/* SPDX-License-Identifier: GPL-2.0 WITH Linux-syscall-note */
```

Reply:
Can you also write down the file path where you note that the defines come from Linux. Maybe in brackets the license with syscall note.